io/io/src/TStreamerInfoActions.cxx

```diff
@@ -339,637 +460,165 @@ namespace TStreamerInfoActions
-      TStreamerElement * aElement  = (TStreamerElement*) config->fCompInfo->fElem;
+      TClass *cle = config->fCompInfo->fClass;
       TMemberStreamer *pstreamer = config->fCompInfo->fStreamer;
-      //TVirtualCollectionProxy *proxy = cl->GetCollectionProxy();
-      //TClass* vClass = proxy ? proxy->GetValueClass() : 0;
-
-      UInt_t eoffset = 0; // extra parameter of TStreamerInfo::WriteBufferAux, 0 for all kind of objects writing
-      UInt_t ioffset = eoffset + config->fOffset;
+      TStreamerElement *aElement = (TStreamerElement *)config->fCompInfo->fElem;
+      UInt_t ioffset = config->fOffset;
+      UInt_t start, count;
+      /* Version_t vers = */ buf.ReadVersion(&start, &count, cle);
 
+      // use same method which is used in kSTL
+      buf.ReadFastArray((void **)((char *)addr + ioffset), cle, config->fCompInfo->fLength, kFALSE, pstreamer);
+      buf.CheckByteCount(start, count, aElement->GetFullName());
+      return 0;
+   }
 
-      UInt_t start,count;
-      Version_t vers = buf.ReadVersion(&start, &count, cle);
+   class TConfWithFactor : public TConfiguration {
+      // Configuration object for the Float16/Double32 where a factor has been specified.
+   public:
+      Double_t fFactor;
+      Double_t fXmin;
+      TConfWithFactor(TVirtualStreamerInfo *info, UInt_t id, TCompInfo_t *compinfo, Int_t offset, Double_t factor, Double_t xmin) : TConfiguration(info,id,compinfo,offset),fFactor(factor),fXmin(xmin) {};
+      TConfiguration *Copy() override { return new TConfWithFactor(*this); }
+   };
 
-      if (!kIsTextT && (vers & TBufferFile::kStreamedMemberWise) ) {
-         // Collection was saved member-wise
+   template <typename T>
+   INLINE_TEMPLATE_ARGS Int_t ReadBasicType_WithFactor(TBuffer &buf, void *addr, const TConfiguration *config)
+   {
+      // Stream a Float16 or Double32 where a factor has been specified.
+      //a range was specified. We read an integer and convert it back to a double.
 
-         vers &= ~( TBufferFile::kStreamedMemberWise );
+      TConfWithFactor *conf = (TConfWithFactor *)config;
+      buf.ReadWithFactor((T*)( ((char*)addr) + config->fOffset ), conf->fFactor, conf->fXmin);
+      return 0;
+   }
 
-         TClass *newClass = aElement->GetNewClass();
-         TClass *oldClass = aElement->GetClassPointer();
-         if( vers < 9 && newClass && newClass!=oldClass ) {
-            Error( "ReadBuffer", "Unfortunately, version %d of TStreamerInfo (used in %s) did not record enough information to convert a %s%s into a %s.",
-                  vers, buf.GetParent() ? buf.GetParent()->GetName() : "memory/socket", oldClass ? oldClass->GetName() : aElement->GetTypeName(), oldClass ? "" : " (could not find the corresponding TClass)", newClass->GetName() );
-            return 0;
-         }
+   class TConfNoFactor : public TConfiguration {
+      // Configuration object for the Float16/Double32 where a factor has been specified.
```

Comment:
It is not quite the right place (and/or would lead to duplicate information).  The full explanation is at https://github.com/root-project/root/blob/6735f228281554a71ca5a2017de6bdaba5a1ae3d/io/io/src/TBufferFile.cxx#L555  

As other place, maybe we should refer to `TBufferFile::WriteFloat16`